### ISSUE TITLE
Raspberry Pi arm build failing

### ISSUE THREAD

**Issue by githubforsank** [CppMicroServices/CppMicroServices#388]
Compiler Details
```
gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/arm-linux-gnueabihf/8/lto-wrapper
Target: arm-linux-gnueabihf
Configured with: ../src/configure -v --with-pkgversion='Raspbian 8.3.0-6+rpi1' --with-bugurl=file:///usr/share/doc/gcc-8/README.Bugs --enable-languages=c,ada,c++,go,d,fortran,objc,obj-c++ --prefix=/usr --with-gcc-major-version-only --program-suffix=-8 --program-prefix=arm-linux-gnueabihf- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-libitm --disable-libquadmath --disable-libquadmath-support --enable-plugin --with-system-zlib --with-target-system-zlib --enable-objc-gc=auto --enable-multiarch --disable-sjlj-exceptions --with-arch=armv6 --with-fpu=vfp --with-float=hard --disable-werror --enable-checking=release --build=arm-linux-gnueabihf --host=arm-linux-gnueabihf --target=arm-linux-gnueabihf
Thread model: posix
gcc version 8.3.0 (Raspbian 8.3.0-6+rpi1) 
``
```
When trying to build the ResourceCompiler fails with the following error.

```
make
[  1%] Building CXX object tools/rc/CMakeFiles/usResourceCompiler.dir/ResourceCompiler.cpp.o
/home/pi/homeos-cmake/cpp-microservices/tools/rc/ResourceCompiler.cpp: In function ‘Json::Value {anonymous}::AggregateManifestsAndValidate(std::unordered_map<std::__cxx11::basic_string<char>, Json::Value>&)’:
/home/pi/homeos-cmake/cpp-microservices/tools/rc/ResourceCompiler.cpp:275:24: error: ‘null’ is not a member of ‘Json::Value’
       if (Json::Value::null == root[iter.name()])
                        ^~~~
make[2]: *** [tools/rc/CMakeFiles/usResourceCompiler.dir/build.make:63: tools/rc/CMakeFiles/usResourceCompiler.dir/ResourceCompiler.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:91: tools/rc/CMakeFiles/usResourceCompiler.dir/all] Error 2
make: *** [Makefile:130: all] Error 2

```
I need to build it with arm compilers. I want to use cppmicroservices in embedded devices like RPi. Please help

**Comment by michkrom** [comment]
Ditto! and There are 3 more errors on RPi armv7 related to cast-align:
jsoncpp.cpp:2253:70: error: cast from ‘const unsigned char*’ to ‘const Json::Value*’ increases required alignment of target type [-Werror=cast-align]
 const Value& Value::nullRef = reinterpret_cast<const Value&>(kNullRef);

**Comment by michkrom** [comment]
To fix you error change this line to 
 if (Json::Value::nullRef == root[iter.name()])

**Comment by jeffdiclemente** [comment]
@githubforsank try using the development branch.

**Comment by michkrom** [comment]
branch development fails with:
/home/pi/.conan/data/CppMicroServices/3.3.1/keysight/stable/build/d02583e133733dc13d0641c4329549e2070176d0/tools/rc/Reso
urceCompiler.cpp:274:24: error: ‘null’ is not a member of ‘Json::Value’
       if (Json::Value::null == root[iter.name()])
                        ^~~~
make[2]: *** [tools/rc/CMakeFiles/usResourceCompiler.dir/build.make:63: tools/rc/CMakeFiles/usResourceCompiler.dir/Resou
rceCompiler.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/pi/.conan/data/CppMicroServices/3.3.1/keysight/stable/build/d02583e133733dc13d0641c4329549e2070176d0/third_party/j
soncpp.cpp:2253:70: error: cast from ‘const unsigned char*’ to ‘const Json::Value*’ increases required alignment of targ
et type [-Werror=cast-align]
 const Value& Value::nullRef = reinterpret_cast<const Value&>(kNullRef);
                                                                      ^
/home/pi/.conan/data/CppMicroServices/3.3.1/keysight/stable/build/d02583e133733dc13d0641c4329549e2070176d0/third_party/j
soncpp.cpp: In function ‘char* Json::duplicateAndPrefixStringValue(const char*, unsigned int)’:
/home/pi/.conan/data/CppMicroServices/3.3.1/keysight/stable/build/d02583e133733dc13d0641c4329549e2070176d0/third_party/j
soncpp.cpp:2334:41: error: cast from ‘char*’ to ‘unsigned int*’ increases required alignment of target type [-Werror=cas
t-align]
   *reinterpret_cast<unsigned*>(newString) = length;
                                         ^
/home/pi/.conan/data/CppMicroServices/3.3.1/keysight/stable/build/d02583e133733dc13d0641c4329549e2070176d0/third_party/j
soncpp.cpp: In function ‘void Json::decodePrefixedString(bool, const char*, unsigned int*, const char**)’:
/home/pi/.conan/data/CppMicroServices/3.3.1/keysight/stable/build/d02583e133733dc13d0641c4329549e2070176d0/third_party/j
soncpp.cpp:2347:58: error: cast from ‘const char*’ to ‘const unsigned int*’ increases required alignment of target type
[-Werror=cast-align]
     *length = *reinterpret_cast<unsigned const*>(prefixed);

due to 1) -Werror -Wcast-align; 2) Json::Value::null should be nullRef.

**Comment by ameya7295** [comment]
Hi @michkrom, I was able to go ahead of the errors you stated by removing the -Wcast-align flag from the CMakeList.txt:406 (not a good practice). 

Going ahead I got a linking error 

`[ 48%] Linking CXX shared library ../lib/libCppMicroServicesd.so
/usr/bin/ld: CMakeFiles/CppMicroServices.dir/src/service/ServiceListeners.cpp.o: in function `std::__atomic_base<unsigned long long>::operator++()':
/usr/include/c++/8/bits/atomic_base.h:296: undefined reference to `__atomic_fetch_add_8'
collect2: error: ld returned 1 exit status
make[2]: *** [framework/CMakeFiles/CppMicroServices.dir/build.make:910: lib/libCppMicroServicesd.so.4.0.0] Error 1
make[1]: *** [CMakeFiles/Makefile2:313: framework/CMakeFiles/CppMicroServices.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
`
@jeffdiclemente would you like to add your comments and some help on this issue.

**Comment by jeffdiclemente** [comment]
@ameya7295 does pulling in the changes from PR #401 resolve the linker error?

**Comment by ameya7295** [comment]
@jeffdiclemente PR #401 is not yet merged with the development branch I see #383 as the last commit to CMakeList.txt. Anyway, I have added the changes by @michkrom to my CMakeList.txt and they work pretty well. The linker issue is solved but getting an error at 

`/CppMicroServices/compendium/tools/SCRCodeGen/Util.hpp:55:22: error: ‘null’ is not a member of ‘Json::Value’
     if (Json::Value::null == jsonVal) {
                      ^~~~
/home/pi/tools/CppMicroServices/compendium/tools/SCRCodeGen/Util.hpp: In constructor ‘codegen::util::JsonValueValidator::JsonValueValidator(const Json::Value&, std::__cxx11::string, codegen::util::JsonValueValidator::ValidChoices<S>&)’:
/home/pi/tools/CppMicroServices/compendium/tools/SCRCodeGen/Util.hpp:80:22: error: ‘null’ is not a member of ‘Json::Value’
     if (Json::Value::null == jsonVal) {
                      ^~~~
make[2]: *** [compendium/tools/SCRCodeGen/CMakeFiles/SCRCodeGen.dir/build.make:76: compendium/tools/SCRCodeGen/CMakeFiles/SCRCodeGen.dir/Main.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:870: compendium/tools/SCRCodeGen/CMakeFiles/SCRCodeGen.dir/all] Error 2
make: *** [Makefile:130: all] Error 2`  

My Environment configuration is:

Hardware : Raspberry Pi B+
OS - Raspbian buster
gcc - 8.3

**Comment by ameya7295** [comment]
Hi @jeffdiclemente , I changed the` Json::Value::null` to `Json::Value::nullRef` in Utils.hpp and it works.
Build success.

**Comment by githubforsank** [comment]
> Ditto! and There are 3 more errors on RPi armv7 related to cast-align:
> jsoncpp.cpp:2253:70: error: cast from ‘const unsigned char*’ to ‘const Json::Value*’ increases required alignment of target type [-Werror=cast-align]
> const Value& Value::nullRef = reinterpret_cast<const Value&>(kNullRef);

You are right .. I also temporarily got it working by removing the cast-align.
This is the pretty obvious fix of nullRef. Thank you :)
And for linking I had to add 
-latomic -l:libatomic.so.1
in the toolchain linker arguments. 

Somehow the cross compile linker doesn't recoginze the implicit need of the libraries.
Is the fix gonna be pull in master ?

**Comment by githubforsank** [comment]
Buster Pi B+ and Pi Zero . Building .. can someone shed light on the use of the host usResourceCompiler being used. I don't understand .cmake imports the binary usResourceCompiler to the current project and then uses it to build bundles. What i am confused at is why can't be the resource compiler also be built and used by the cross compiler

**Comment by michkrom** [comment]
There is a pending pull request with fixes. When I get the power back  I
will finish it.

**Comment by jeffdiclemente** [comment]
@githubforsank 
> Buster Pi B+ and Pi Zero . Building .. can someone shed light on the use of the host usResourceCompiler being used. I don't understand .cmake imports the binary usResourceCompiler of the host to the current build and then uses it to build bundles. What i am confused at is why can't be the resource compiler also be built and used by the cross compiler

I'm not sure I understand what the issue is. usResourceCompiler is a build tool used to embed meta-data into the bundle. It can be used during cross compilation. The build host usResourceCompiler is used in the cross compilation use case. 

Can you elaborate on what you are seeing?

**Comment by jeffdiclemente** [comment]
Fixed in PR #401
